### ISSUE TITLE
Alternative way of detecting cluster changes instead of `kubeconfig`

### ISSUE THREAD

**Issue by eplightning** [pulumi/pulumi-kubernetes#2745]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently the provider relies on passed `kubeconfig` input to decide if it needs replacement or not.
This is not always correct and has a chance for both false negatives and false positives, resulting either in potentially destructive replacement of resources (volumes, ArgoCD applications...) or errors when an expected resource doesn't exist.

Therefore I'd like to request an alternative to doing that for users that have this information by other means (for example ID of a managed K8s cluster).

Simplest way I can think of to achieve this would be adding a new optional provider config property `clusterId` and changing the detection logic as follows:
1. If `clusterId` is passed, report provider for replacement whenever new `clusterId` is computed or differs from previous `clusterId`
2. Else use current kubeconfig logic

Then this could be used this way in Pulumi code:
```typescript
const aks = new containerservice.KubernetesCluster('cluster', ...);
const k8sProvider = new kubernetes.Provider('cluster', {
  clusterId: aks.id,
  kubeconfig: aks.kubeAdminConfigRaw
});
```

### Affected area/feature

Provider

**Comment by EronWright** [comment]
Another alternative may be to use some server information, such as the UID of the default namespace.  I don't know whether the provider resource could do the detection, and vary its URN (identity) accordingly.  Or whether each object resource would do the detection and state tracking.

**Comment by EronWright** [comment]
I think we should prioritize a simple fix that gives the user control, like a `clusterid` property that causes replacement whenever it changes but has no other significance, and fall back to existing behavior if `clusterid` is not set.

**Comment by blampe** [comment]
> I think we should prioritize a simple fix that gives the user control, like a `clusterid` property that causes replacement whenever it changes but has no other significance, and fall back to existing behavior if `clusterid` is not set.

Agree!

Roughly speaking, I think we want `DiffConfig` to have behavior like:

* Setting or unsetting a clusterid on an existing provider → update the provider.
* Other config changes if clusterid is set and unchanged, or being set/unset → update the provider.
* Changing a non-null clusterid to a different non-null clusterid → replace the provider.

Otherwise fall back to existing `DiffConfig` logic when clusterid is unset.

> Another alternative may be to use some server information, such as the UID of the default namespace. I don't know whether the provider resource could do the detection, and vary its URN (identity) accordingly. Or whether each object resource would do the detection and state tracking.

I really like this idea but I worry there's an edge case where the user might not have access to the default namespace. It's not uncommon to restrict access to default so it doesn't become a dumping ground, although in practice I've only seen this done with admission webhooks.

**Comment by blampe** [comment]
FYI @eplightning et al. -- `clusterIdentifier` provider configuration is available in `4.16.0-alpha.1721342660`. Once a provider has `clusterIdentifier`, subsequent changes to the provider will _always_ be updates _unless_ the `clusterIdentifier` changes.

Give it a spin and let us know how it works for you!

**Comment by andrewstuart** [comment]
Is there a way to use this in the Go language environment? From what I can see, the supporting code is committed, but it still wants to load a plugin version `v4.0.0-alpha.0+dev`, since it looks like the release process is what actually sets the right versions in various files, meaning we probably need a version cut to try this fix, as the issue is currently affecting us.

**Comment by loveRhythm1990** [comment]
when will a new tag will be released? we also encountered this problem and really hope a new tag, @blampe

**Comment by blampe** [comment]
@loveRhythm1990 this is available in v4.16.0. Pre-release alphas are also always available, e.g. https://www.npmjs.com/package/@pulumi/kubernetes?activeTab=versions
